README.md
libchardet - Mozilla's Universal Charset Detector C/C++ API
===
[![Build Status](https://travis-ci.org/Joungkyun/libchardet.svg?branch=master)](https://travis-ci.org/Joungkyun/libchardet)
[![GitHub license](https://img.shields.io/badge/license-MPL%201.1-blue.svg)](https://raw.githubusercontent.com/Joungkyun/libchardet/master/LICENSE)
[![GitHub last release](https://img.shields.io/github/release/joungkyun/libchardet.svg)](https://github.com/joungkyun/libchardet/releases)
[![GitHub closed issues](https://img.shields.io/github/issues-closed-raw/joungkyun/libchardet.svg)](https://github.com/joungkyun/libchardet/issues?q=is%3Aissue+is%3Aclosed)
[![GitHub closed pull requests](https://img.shields.io/github/issues-pr-closed-raw/joungkyun/libchardet.svg)](https://github.com/joungkyun/libchardet/pulls?q=is%3Apr+is%3Aclosed)

## License
Copyright (c) 2019 JoungKyun.Kim <http://oops.org> All rights reserved.

This program is under MPL 1.1 or LGPL 2.1

## Description
libchardet is based on Mozilla Universal Charset Detector library and, detects
the character set used to encode data.

[Original code](http://lxr.mozilla.org/seamonkey/source/extensions/universalchardet/) was writed by Netscape Communications Corporation, Techniques used by universalchardet are described at &lt;http://www-archive.mozilla.org/projects/intl/UniversalCharsetDetection.html&gt;.

libchardet see also John Gardiner Myers's [Encode-Detect-1.01](http://search.cpan.org/~jgmyers/Encode-Detect-1.01/)
perl module, and added C wrapping API, and library build environment with libtool.

From 1.0.5, libchardet was reflected single-byte charset detection confidence
algorithm of [uchardet](https://github.com/BYVoid/uchardet/) and new language models.
(Arabic, Danish, Esperanto, German, Spanish, Turkish, Vietnamese)

From 1.0.6, bom members have been added to the DetectObj structure.
The value of the bom member is 1, which means that it has been detected as a BOM.
Support for bom member can be determined by the existence of the CHARDET_BOM_CHECK
constant. See example below.

## Installation

See also [INSTALL](INSTALL) document

## Sample Codes

See also test directory of source code

```c
       #include <chardet.h>

       int main (void) {
            DetectObj *obj;
            char * str = "안녕하세요";

            if ( (obj = detect_obj_init ()) == NULL ) {
                 fprintf (stderr, "Memory Allocation failed\n");
                 return CHARDET_MEM_ALLOCATED_FAIL;
            }

       #ifndef CHARDET_BINARY_SAFE 
            // before 1.0.5. This API is deprecated on 1.0.5
            switch (detect (str, &obj))
       #else
            // from 1.0.5
            switch (detect_r (str, strlen (str), &obj))
       #endif
            {
                 case CHARDET_OUT_OF_MEMORY :
                      fprintf (stderr, "On handle processing, occured out of memory\n");
                      detect_obj_free (&obj);
                      return CHARDET_OUT_OF_MEMORY;
                 case CHARDET_NULL_OBJECT :
                      fprintf (stderr,
                                "2st argument of chardet() is must memory allocation "
                                "with detect_obj_init API\n");
                      return CHARDET_NULL_OBJECT;
            }

        #ifndef CHARDET_BOM_CHECK
            printf ("encoding: %s, confidence: %f\n", obj->encoding, obj->confidence);
        #else
            // from 1.0.6 support return whether exists BOM
            printf (
                "encoding: %s, confidence: %f, exist BOM: %d\n",
                 obj->encoding, obj->confidence, obj->bom
            );
        #endif
            detect_obj_free (&obj);

            return 0;
       }
```

or looping code

```c
       #include <chardet.h>

       int main (void) {
            Detect    * d;
            DetectObj * obj;
            char * str = "안녕하세요";

            if ( (d = detect_init ()) == NULL ) {
                 fprintf (stderr, "chardet handle initialize failed\n");
                 return CHARDET_MEM_ALLOCATED_FAIL;
            }

            while ( 1 ) {
                detect_reset (&d);

                if ( (obj = detect_obj_init ()) == NULL ) {
                     fprintf (stderr, "Memory Allocation failed\n");
                     return CHARDET_MEM_ALLOCATED_FAIL;
                }

       #ifndef CHARDET_BINARY_SAFE 
                // before 1.0.5. This API is deprecated on 1.0.5
                switch (detect_handledata (&d, str,, &obj))
       #else
                // from 1.0.5
                switch (detect_handledata_r (&d, str, strlen (str), &obj))
       #endif
                {
                     case CHARDET_OUT_OF_MEMORY :
                          fprintf (stderr, "On handle processing, occured out of memory\n");
                          detect_obj_free (&obj);
                          return CHARDET_OUT_OF_MEMORY;
                     case CHARDET_NULL_OBJECT :
                          fprintf (stderr,
                                    "2st argument of chardet() is must memory allocation "
                                    "with detect_obj_init API\n");
                          return CHARDET_NULL_OBJECT;
                }

        #ifndef CHARDET_BOM_CHECK
                printf ("encoding: %s, confidence: %f\n", obj->encoding, obj->confidence);
        #else
                // from 1.0.6 support return whether exists BOM
                printf (
                    "encoding: %s, confidence: %f, exist BOM: %d\n",
                    obj->encoding, obj->confidence, obj->bom
                );
        #endif
                detect_obj_free (&obj);

                if ( 1 )
                    break;
            }
            detect_destroy (&d);

           return 0;
       }
```

## APIs
  * PHP Extension    - https://github.com/OOPS-ORG-PHP/mod_chardet
  * PYTHON C Binding - https://github.com/Joungkyun/python-chardet
  * PERL             - http://search.cpan.org/~jgmyers/Encode-Detect-1.01/Detect.pm


Changelog
Mozilla's Universal Charset Detector C/C++ library

*** 1.0.7

2021/05/25 KST
	- give a little weight to "probable sequences" about single byte charset
	- update langage table for Greek, Thai and Vietnamese


*** 1.0.6

2021/05/14 KST
	- fixed #14 can't detect short euc-kr

2021/05/13 KST
	- fixed #18 SECURITY! Invalid memory approach (heap-use-after-free) (@gaoxiang-ut)

2019/08/01 KST
	- fixed #13 bom member has been added to the DetectObj structure
	- fixed #15 support automake style 'make check'

2017/03/07 KST
	- fixed #12 No include guard

2016/08/20 KST
	- fixed #9 configure.ac needs subdir-objects
	- fixed #10 autogen failure because AM_PROG_AR with automake 1.11.1


*** 1.0.5

2016/05/11 KST
	- release 1.0.5
	- fixed #6 can not detect EUC extended area

2016/05/05 KST
	- fixed #5 no detect control character on US-ASCII
	- fixed #7 wrong detect Danash ISO-8859-15
	- fixed #8 fixed binary safe problems
	  . replace detect to detect_r
	  . replace detect_handledata to detect_handledata_r

2016/05/04 KST
	- fixed #1 separate model directory
	- merge uchardet's improvement
	  . fixed #2 Improve single-byte charset detection confidence algorithm
	  . fixed #3 update model of Greek, Hungarian and Thai
	  . fixed #4 new language model (Arabic, Danish, Esperanto, German, Spanish,
	             Turkish, Vietnamese)

2015/12/10 KST
	- fixed man pages wrong macro bug (martin.gansser@gmail.com)


*** 1.0.4

2014/02/14 KST
	- fixed duplicated path on chardet.pc
	- release 1.0.4

2014/02/13 KST
	- support windows native library with MinGW
	- support Code::Blocks project

2014/02/11 KST
	- fixed build error and shared library on cygwin


*** 1.0.3

2014/02/10 KST
	- add chardet.pc by Lee ByungYoung <darklin20@gamil.com>
	- applied automake
	- add english man page

2012/08/14 KST
	- fixed comparison on JpCntx.cpp


*** 1.0.2

2010/07/05 KST
	- support visibility attribute on gcc4.
	- add version information api


*** 1.0.1

2009/02/23 KST
	- fix wrong detect TIS-620 charset
	- release 1.0.1


*** 1.0.0

2009/02/23 KST JoungKyun.Kim <http://oops.org>
	- add version constant

2009/02/22 KST JoungKyun.Kim <http://oops.org>
	- release 1.0.0 based on perl Encode-Detect-1.01



src/chardet.h
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * Mozilla's universal charset detector C/C++ Wrapping API
 *      Writer(s) :
 *          Detect class by John Gardiner Myers <jgmyers@proofpoint.com>
 *          C wrapping API by JoungKyun.Kim <http://oops.org>
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

#ifndef CHARDET_H
#define CHARDET_H

/*
#if defined _WIN32 || defined __CYGWIN__
	#define CHARDET_API
#else
	#if defined(__GNUC__) && __GNUC__ >= 4
		#define CHARDET_API __attribute__ ((visibility("default")))
	#else
		#define CHARDET_API
	#endif
#endif
*/


#ifdef HAVE_CONFIG_H
#include <chardet-config.h>
#endif

#if defined _WIN32 || defined __CYGWIN__
	#ifdef HAVE_DLL_EXPORT
		#define CHARDET_API __declspec(dllexport)
	#else
		#define CHARDET_API __declspec(dllimport)
	#endif
#else
	#ifdef HAVE_VISIBILITY
		#define CHARDET_API __attribute__ ((visibility("default")))
	#else
		#define CHARDET_API
	#endif
#endif

#include <version.h>

#include <stdio.h>
#include <string.h>

#define CHARDET_OUT_OF_MEMORY -128
#define CHARDET_MEM_ALLOCATED_FAIL -127

#define CHARDET_SUCCESS     0
#define CHARDET_NO_RESULT   1
#define CHARDET_NULL_OBJECT 2

// whether to support detect_r and detect_handledata_r API
#define CHARDET_BINARY_SAFE 1

// whether to support bom member of DetectObj structure
#define CHARDET_BOM_CHECK 1

#ifdef __cplusplus
extern "C" {
#endif
	typedef struct Detect_t Detect;

	typedef struct DetectObject {
		char * encoding;
		float confidence;
		short bom;
	} DetectObj;

	CHARDET_API char * detect_version (void);
	CHARDET_API char * detect_uversion (void);

	CHARDET_API DetectObj * detect_obj_init (void);
	CHARDET_API void detect_obj_free (DetectObj **);

	CHARDET_API Detect * detect_init (void);
	CHARDET_API void detect_reset (Detect **);
	CHARDET_API void detect_dataend (Detect **);
	CHARDET_API short detect_handledata (Detect **, const char *, DetectObj **);
	CHARDET_API short detect_handledata_r (Detect **, const char *, size_t, DetectObj **);
	CHARDET_API void detect_destroy (Detect **);
	CHARDET_API short detect (const char *, DetectObj **);
	CHARDET_API short detect_r (const char *, size_t, DetectObj **);
#ifdef __cplusplus
};
#endif

#endif // close define CHARDET_H

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */
